ir/src/depgraph.rs

```diff
@@ -0,0 +1,433 @@
+        assert!(dir.is_dir());
+        if dirs_visited.contains(dir) {
+            return None;
+        }
+
+        let mut frontier = vec![dir.to_owned()];
+        let mut results: Option<Vec<PathBuf>> = None;
+
+        while let Some(dir) = frontier.pop() {
+            let dir_entries = fs::read_dir(&dir).expect("Unable to iterate directory");
+            for dir_entry in dir_entries {
+                let dir_entry = dir_entry.expect("Cannot read dir entry").path();
+
+                // Queue subdir for processing. Bravely assume lack of cycles.
+                if dir_entry.is_dir() && !dirs_visited.contains(&dir_entry) {
+                    frontier.push(dir_entry.to_owned());
+                }
+
+                // A path we've never seen before?!
+                if !self.entries.contains_key(&dir_entry) {
+                    let new_files = results.get_or_insert_with(Vec::new);
+                    new_files.push(dir_entry);
+                }
+            }
+            dirs_visited.insert(dir.to_owned());
+        }
+        results
+    }
+
+    /// Set of things that have either changed themselves or had
+    /// something they depend on change.
+    pub fn changed(&self) -> Result<Vec<Change>, io::Error> {
+        let mut paths: Vec<&PathBuf> = self.entries.iter().map(|(k, _)| k).collect();
+
+        let mut changes: Vec<Change> = Vec::new();
+
+        // Add any new files in tracked dirs
+        let mut dirs_visited: HashSet<PathBuf> = HashSet::new();
+        let new_paths: Vec<PathBuf> = paths
+            .iter()
+            .filter(|p| p.is_dir())
+            .flat_map(|p| self.new_paths(&mut dirs_visited, p).unwrap_or_default())
+            .collect();
+        new_paths.iter().for_each(|p| paths.push(p));
+
+        // Anything change 'round here?
+        for path in paths {
+            let updated_entry = DepGraphEntry::new(path)?;
+            let has_changed = if self.entries.contains_key(path) {
+                let maybe_current = self.entries.get(path).unwrap();
+                maybe_current
+                    .as_ref()
+                    .map(|current_entry| updated_entry != *current_entry)
+                    .unwrap_or(true)
+            } else {
+                true
+            };
+
+            if has_changed {
+                changes.push(Change {
+                    path: path.to_owned(),
+                    updated_entry,
+                });
+            }
+        }
+
+        Ok(changes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{fs, ops::Add, path::PathBuf, time::Duration};
+
+    use filetime::set_file_mtime;
+    use tempfile::{tempdir, TempDir};
+
+    use super::DepGraph;
+
+    fn assert_no_changes(dg: &DepGraph) {
+        assert_eq!(
+            Vec::<&PathBuf>::new(),
+            dg.changed()
+                .unwrap()
+                .iter()
+                .map(|c| &c.path)
+                .collect::<Vec<&PathBuf>>()
+        )
+    }
```

Comment:
this logic is confusing, why not just,

```rust
assert_eq!(db.changed().unwrap().iter().count(), 0);
```

is it that you want to see the entries if the assert fails?

Reply:
I specifically want to see the mismatched entries on failure